akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileAndResourceDirectives.scala

```diff
@@ -127,14 +149,26 @@ trait FileAndResourceDirectives {
    *
    * @group fileandresource
    */
-  def getFromDirectory(directoryName: String)(implicit resolver: ContentTypeResolver): Route = {
-    val base = withTrailingSlash(directoryName)
-    extractUnmatchedPath { path ⇒
+  def getFromDirectory(directoryName: String)(implicit resolver: ContentTypeResolver): Route =
+    getFromDirectory(Paths.get(withTrailingSlash(directoryName)))(resolver)
+
+  /**
+   * Completes GET requests with the content of a file underneath the given directory.
+   * If the file cannot be read the Route rejects the request.
+   *
+   * @group fileandresource
+   */
+  def getFromDirectory(path: Path)(implicit resolver: ContentTypeResolver): Route = {
+    val base = path.toRealPath() // follow links
+    extractUnmatchedPath { unmatchedPath ⇒
       extractLog { log ⇒
-        fileSystemPath(base, path, log) match {
-          case ""       ⇒ reject
-          case fileName ⇒ getFromFile(fileName)
-        }
+        val path = if (unmatchedPath.startsWithSlash) unmatchedPath.tail else unmatchedPath
+        val file = base.resolve(path.toString)
+        if (exists(file, LinkOption.NOFOLLOW_LINKS) && file.toRealPath().startsWith(base)) // FIXME: optional symlinks
+          // TODO log
```

Comment:
Good point. We could alternatively just log paths that are outside the base directory.